doc/memos/rdm-draft-alamos-lanzieri-runtime-configuration-architecture.md

```diff
@@ -0,0 +1,471 @@
+required to migrate the data between storage facilities (e.g to migrate all
+configurations from SF A to B, register B as source and destination and add
+A as a source).
+
+A conceptual example of a SF can be found in the [Appendix](#Appendix).
+
+## 3.3. RIOT Registry Usage Flow
+
+### Registry Initialization
+Modules declare and register **RH** for configuration groups
+in the RIOT Registry. **Storage facilities** are registered as sources and/or
+destinations of configurations in the RIOT Registry.
+
+<img align="center" src="./files/rdm-draft-alamos-lanzieri-runtime-configuration-architecture/registry-boot.svg" width="300" />
+
+### Get, set, apply and export configurations
+At any time, the application or a configuration manager can _retrieve_ a
+configuration value (`registry_get_value`), _set_ a configuration value
+(`registry_set_value`), _apply_ configuration changes (`registry_commit`) or
+_export_ configurations using a user-defined callback function
+(`registry_export`).
+
+Note these functions don't interact with SF, so
+configuration changes are not reflected in the non-volatile storage devices
+unless `registry_store` is called (see [Load and store configurations](#load-and-store-configurations))
+
+The following diagram shows the process of each function. It's assumed there
+are 2 RH registered in the RIOT Registry: a _cord_ configuration
+group with a Resource Directory Server IP Address (`rd_ip_addr`) and an
+_Application_ configuration group with a `foo` configuration parameter.
+
+<img src="./files/rdm-draft-alamos-lanzieri-runtime-configuration-architecture/registry-rh.svg" />
+
+### Load and store configurations
+At any time, the application or a configuration manager can _load_ all
+configurations from all SF sources (`registry_load` function) or
+_store_ them in the SF destination (`registry_store` function).
+
+As one could expect, `registry_load` will call the SF `load`
+handler with `registry_set_value` as callback. In the a similar way,
+`registry_store` will navigate through all RH and call their
+_export_ function with the SF _store_ handler as callback.  
+
+The following diagram shows these processes:
+
+<img src="./files/rdm-draft-alamos-lanzieri-runtime-configuration-architecture/registry-storage.svg" />
+
+# 4. Configuration managers
+Configuration managers are modules that allow a RIOT node to be
+configured from one or more communication interfaces. Examples of these
+communication interfaces could be UART, SPI or higher layers like PPP, IPv6,
+UDP, CoAP, etc.
+
+These are some examples of Configuration Managers:
+- [OMA LWM2M](https://www.omaspecworks.org/what-is-oma-specworks/iot/lightweight-m2m-lwm2m/)
+- [Newt Manager](https://mynewt.apache.org/master/os/modules/devmgmt/newtmgr.html)
+  via interfaces like NFC, BLE, Serial, etc.
+- A UART shell with special commands for interacting with configurations (CLI)
+
+A Configuration Manager may interact with the RIOT Registry to access Runtime
+Configurations.
```

Comment:
Is this really a "MAY" here? According to the first architecture graph, the Configuration Manager accesses the RIOT Registry. 
It would be IMO helpful to clarify 
- which is the interface or surface the Configuration Manager sees and interacts with
- which (southbound) API(s) does it need to implement towards RIOT

Reply:
> Is this really a "MAY" here? According to the first architecture graph, the Configuration Manager accesses the RIOT Registry.

You are right, the Configuration Managers should only access the parameters via the Registry.

> It would be IMO helpful to clarify
> 
> * which is the interface or surface the Configuration Manager sees and interacts with
> * which (southbound) API(s) does it need to implement towards RIOT

Each implemented Configuration Manager will interact directly with the RIOT Registry API.

Regarding the API of the Configuration Managers towards RIOT, I'm not sure if we extract a common interface it at the moment, as all of them are really different  (IMO Configuration Managers could even require an RDM on their own). We can though define some responsibilities they have (e.g. access control mechanisms, enable/disable of interfaces). Maybe that can be rephrased in the document.